src/base/locked_queue.h

```diff
@@ -61,56 +65,111 @@ class LockedQueue {
       index_.Put(queue_.insert(queue_.end(), std::move(obj)), shard);
     }
     ++size_;
-    pop_condition_.notify_one();
+    index_.NotifyShard(shard);
+
+    // See a comment on |aggregator_condition_| to see why two notifications
+    // should be done here.
+    aggregator_condition_.notify_one();
 
     return true;
   }
 
   // Returns disengaged object only when this queue is closed and empty.
-  Optional Pop(ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
+  Optional Pop() THREAD_SAFE {
     UniqueLock lock(pop_mutex_);
-    pop_condition_.wait(lock, [this] { return closed_ || !queue_.empty(); });
+    index_.WaitShard(DEFAULT_SHARD, lock, [this] {
+      return closed_ || !queue_.empty();
+    });
     if (closed_ && queue_.empty()) {
       return Optional();
     }
 
-    auto it = index_.Get(shard, queue_.begin());
-    Optional&& obj = std::move(*it);
-    queue_.erase(it);
-    --size_;
-    return std::move(obj);
+    return RemoveTaskFromQueue(
+        index_.GetWithHint(DEFAULT_SHARD, queue_.begin()));
   }
 
   // Returns disengaged object only when this queue is closed and empty, or pool
   // is shutting down.
-  Optional Pop(const WorkerPool& pool, ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
+  Optional Pop(const WorkerPool& pool,
+               const ui32 shard_queue_limit = NOT_STRICT_SHARDING,
+               const ui32 shard = DEFAULT_SHARD) THREAD_SAFE {
     DCHECK(timeout_ > Seconds::zero());
 
     UniqueLock lock(pop_mutex_);
-    do {
-      pop_condition_.wait_for(lock, timeout_);
-    } while (!closed_ && queue_.empty() && !pool.IsShuttingDown());
+    if (shard_queue_limit == NOT_STRICT_SHARDING) {
+      return PopWithHint(pool, lock, shard);
+    } else {
+      return PopStrict(pool, lock, shard_queue_limit, shard);
+    }
+  }
+
+ private:
+  Optional PopWithHint(const WorkerPool& pool,
+                       UniqueLock& lock,
+                       const ui32 shard) THREAD_SAFE {
+    // One can't wait for condition using predicate and timed wait here as once
+    // the waiting timed out, the |pool.IsShuttingDown()| should be checked and
+    // wait again if pool doesn't shutting down.
+    while (!closed_ && queue_.empty() && !pool.IsShuttingDown()) {
+      index_.WaitShardFor(shard, lock, timeout_);
+    }
     if ((closed_ && queue_.empty()) || pool.IsShuttingDown()) {
       return Optional();
     }
+    return RemoveTaskFromQueue(index_.GetWithHint(shard, queue_.begin()));
```

Comment:
You should mark the method `PopWithHint` as `THREAD_UNSAFE` otherwise it won't work in multiple threads.